File: step_09_stack_queues/9.1_learning/P001_implement_stack_using_arrays/Solution.java
/**
 * Problem: Implement Stack using Arrays
 * Difficulty: EASY | XP: 10
 *
 * Implement a stack with push, pop, peek, isEmpty, size using an array.
 *
 * @author DSA_Nova
 */
public class Solution {

    // ============================================================
    // STACK IMPLEMENTATION (Fixed-Size Array)
    // All operations: O(1) time, O(1) space per operation
    // Overall space: O(capacity)
    // ============================================================
    static class Stack {
        private int[] arr;
        private int top;
        private int capacity;

        Stack(int capacity) {
            this.capacity = capacity;
            this.arr = new int[capacity];
            this.top = -1;  // -1 means empty
        }

        /**
         * Push element onto the stack.
         * Time: O(1)
         */
        void push(int val) {
            if (top == capacity - 1) {
                System.out.println("Stack Overflow! Cannot push " + val);
                return;
            }
            top++;
            arr[top] = val;
        }

        /**
         * Remove and return the top element.
         * Time: O(1)
         */
        int pop() {
            if (top == -1) {
                System.out.println("Stack Underflow! Cannot pop.");
                return -1;
            }
            int val = arr[top];
            top--;
            return val;
        }

        /**
         * Return the top element without removing it.
         * Time: O(1)
         */
        int peek() {
            if (top == -1) {
                System.out.println("Stack is empty! Cannot peek.");
                return -1;
            }
            return arr[top];
        }

        /**
         * Check if stack is empty.
         * Time: O(1)
         */
        boolean isEmpty() {
            return top == -1;
        }

        /**
         * Return number of elements in the stack.
         * Time: O(1)
         */
        int size() {
            return top + 1;
        }

        /**
         * Print stack contents (bottom to top).
         */
        void printStack() {
            if (isEmpty()) {
                System.out.println("Stack: (empty)");
                return;
            }
            StringBuilder sb = new StringBuilder("Stack (bottom->top): [");
            for (int i = 0; i <= top; i++) {
                sb.append(arr[i]);
                if (i < top) sb.append(", ");
            }
            sb.append("]");
            System.out.println(sb.toString());
        }
    }

    // ============================================================
    // DYNAMIC STACK (with resizing) - Amortized O(1)
    // ============================================================
    static class DynamicStack {
        private int[] arr;
        private int top;
        private int capacity;

        DynamicStack() {
            this.capacity = 4;  // start small
            this.arr = new int[capacity];
            this.top = -1;
        }

        void push(int val) {
            if (top == capacity - 1) {
                resize(capacity * 2);
            }
            arr[++top] = val;
        }

        int pop() {
            if (top == -1) return -1;
            return arr[top--];
        }

        int peek() {
            if (top == -1) return -1;
            return arr[top];
        }

        boolean isEmpty() {
            return top == -1;
        }

        int size() {
            return top + 1;
        }

        private void resize(int newCapacity) {
            int[] newArr = new int[newCapacity];
            for (int i = 0; i <= top; i++) {
                newArr[i] = arr[i];
            }
            arr = newArr;
            capacity = newCapacity;
            System.out.println("  (Resized to capacity " + newCapacity + ")");
        }
    }

    public static void main(String[] args) {
        System.out.println("=== Implement Stack using Arrays ===\n");

        // Fixed-size stack demo
        System.out.println("--- Fixed-Size Stack (capacity=5) ---");
        Stack stack = new Stack(5);

        stack.push(10);
        stack.push(20);
        stack.push(30);
        stack.printStack();

        System.out.println("Peek: " + stack.peek());     // 30
        System.out.println("Size: " + stack.size());      // 3
        System.out.println("Pop:  " + stack.pop());       // 30
        System.out.println("Pop:  " + stack.pop());       // 20
        stack.printStack();

        System.out.println("isEmpty: " + stack.isEmpty()); // false
        System.out.println("Pop:  " + stack.pop());        // 10
        System.out.println("isEmpty: " + stack.isEmpty()); // true

        // Overflow test
        System.out.println("\n--- Overflow Test ---");
        Stack small = new Stack(2);
        small.push(1);
        small.push(2);
        small.push(3);  // Overflow!

        // Underflow test
        System.out.println("\n--- Underflow Test ---");
        Stack empty = new Stack(3);
        empty.pop();    // Underflow!
        empty.peek();   // Empty!

        // Dynamic stack demo
        System.out.println("\n--- Dynamic Stack (auto-resize) ---");
        DynamicStack dStack = new DynamicStack();
        for (int i = 1; i <= 10; i++) {
            dStack.push(i * 10);
            System.out.println("Pushed " + (i * 10) + ", size=" + dStack.size());
        }
        System.out.println("Peek: " + dStack.peek());     // 100
        System.out.println("Pop:  " + dStack.pop());       // 100
        System.out.println("Size: " + dStack.size());      // 9
    }
}

File: step_09_stack_queues/9.1_learning/P001_implement_stack_using_arrays/solution.py
"""
Problem: Implement Stack using Arrays
Difficulty: EASY | XP: 10

Implement stack with push, pop, peek, isEmpty, size using array.
"""


# ============================================================
# STACK IMPLEMENTATION (Fixed-Size Array)
# All operations: O(1) time, O(1) space per operation
# Overall space: O(capacity)
# ============================================================
class Stack:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.arr = [0] * capacity
        self.top = -1  # -1 means empty

    def push(self, val: int) -> None:
        """Push element onto the stack. Time: O(1)."""
        if self.top == self.capacity - 1:
            print(f"Stack Overflow! Cannot push {val}")
            return
        self.top += 1
        self.arr[self.top] = val

    def pop(self) -> int:
        """Remove and return top element. Time: O(1)."""
        if self.top == -1:
            print("Stack Underflow! Cannot pop.")
            return -1
        val = self.arr[self.top]
        self.top -= 1
        return val

    def peek(self) -> int:
        """Return top element without removing. Time: O(1)."""
        if self.top == -1:
            print("Stack is empty! Cannot peek.")
            return -1
        return self.arr[self.top]

    def is_empty(self) -> bool:
        """Check if stack is empty. Time: O(1)."""
        return self.top == -1

    def size(self) -> int:
        """Return number of elements. Time: O(1)."""
        return self.top + 1

    def print_stack(self) -> None:
        """Print stack contents (bottom to top)."""
        if self.is_empty():
            print("Stack: (empty)")
            return
        elements = [str(self.arr[i]) for i in range(self.top + 1)]
        print(f"Stack (bottom->top): [{', '.join(elements)}]")


# ============================================================
# DYNAMIC STACK (with resizing) - Amortized O(1)
# Python lists already do this, but we implement it explicitly.
# ============================================================
class DynamicStack:
    def __init__(self):
        self.capacity = 4  # start small
        self.arr = [0] * self.capacity
        self.top = -1

    def push(self, val: int) -> None:
        if self.top == self.capacity - 1:
            self._resize(self.capacity * 2)
        self.top += 1
        self.arr[self.top] = val

    def pop(self) -> int:
        if self.top == -1:
            return -1
        val = self.arr[self.top]
        self.top -= 1
        return val

    def peek(self) -> int:
        if self.top == -1:
            return -1
        return self.arr[self.top]

    def is_empty(self) -> bool:
        return self.top == -1

    def size(self) -> int:
        return self.top + 1

    def _resize(self, new_capacity: int) -> None:
        new_arr = [0] * new_capacity
        for i in range(self.top + 1):
            new_arr[i] = self.arr[i]
        self.arr = new_arr
        self.capacity = new_capacity
        print(f"  (Resized to capacity {new_capacity})")


if __name__ == "__main__":
    print("=== Implement Stack using Arrays ===\n")

    # Fixed-size stack demo
    print("--- Fixed-Size Stack (capacity=5) ---")
    stack = Stack(5)

    stack.push(10)
    stack.push(20)
    stack.push(30)
    stack.print_stack()

    print(f"Peek: {stack.peek()}")       # 30
    print(f"Size: {stack.size()}")       # 3
    print(f"Pop:  {stack.pop()}")        # 30
    print(f"Pop:  {stack.pop()}")        # 20
    stack.print_stack()

    print(f"isEmpty: {stack.is_empty()}")  # False
    print(f"Pop:  {stack.pop()}")          # 10
    print(f"isEmpty: {stack.is_empty()}")  # True

    # Overflow test
    print("\n--- Overflow Test ---")
    small = Stack(2)
    small.push(1)
    small.push(2)
    small.push(3)  # Overflow!

    # Underflow test
    print("\n--- Underflow Test ---")
    empty = Stack(3)
    empty.pop()    # Underflow!
    empty.peek()   # Empty!

    # Dynamic stack demo
    print("\n--- Dynamic Stack (auto-resize) ---")
    d_stack = DynamicStack()
    for i in range(1, 11):
        d_stack.push(i * 10)
        print(f"Pushed {i * 10}, size={d_stack.size()}")
    print(f"Peek: {d_stack.peek()}")    # 100
    print(f"Pop:  {d_stack.pop()}")     # 100
    print(f"Size: {d_stack.size()}")    # 9
